File: pkg/material/diffuse.go
package material

import (
	"math"
	"math/rand"

	"sif/gogs/eric/relray/pkg/geometry"
	"sif/gogs/eric/relray/pkg/spectrum"
	"sif/gogs/eric/relray/pkg/vec"
)

// Diffuse is a Lambertian diffuse material with spectral reflectance.
type Diffuse struct {
	Reflectance spectrum.SPD
}

func (d *Diffuse) Scatter(inDir vec.Vec3, hit geometry.Hit) ScatterResult {
	// Cosine-weighted hemisphere sampling via normal + random unit vector
	scattered := hit.Normal.Add(randomUnitVec())
	// Catch degenerate case where random vector cancels normal
	if scattered.LengthSq() < 1e-12 {
		scattered = hit.Normal
	}
	return ScatterResult{
		Scattered:   true,
		OutDir:      scattered.Normalize(),
		Reflectance: d.Reflectance,
	}
}

func (d *Diffuse) Emitted(hit geometry.Hit) spectrum.SPD {
	return spectrum.SPD{}
}

func randomUnitVec() vec.Vec3 {
	for {
		v := vec.Vec3{
			X: 2*rand.Float64() - 1,
			Y: 2*rand.Float64() - 1,
			Z: 2*rand.Float64() - 1,
		}
		l2 := v.LengthSq()
		if l2 > 1e-6 && l2 <= 1 {
			return v.Scale(1.0 / math.Sqrt(l2))
		}
	}
}

File: rust/src/material/diffuse.rs
use crate::geometry::Hit;
use crate::spectrum::Spd;
use crate::vec::Vec3;
use rand::rngs::SmallRng;

use super::{random_unit_vec, Material, ScatterResult};

/// Lambertian diffuse material with spectral reflectance.
pub struct Diffuse {
    pub reflectance: Spd,
}

impl Material for Diffuse {
    fn scatter(&self, _in_dir: Vec3, hit: &Hit, rng: &mut SmallRng) -> ScatterResult {
        let mut scattered = hit.normal + random_unit_vec(rng);
        if scattered.length_sq() < 1e-12 {
            scattered = hit.normal;
        }
        ScatterResult {
            scattered: true,
            out_dir: scattered.normalize(),
            reflectance: self.reflectance,
        }
    }

    fn emitted(&self, _hit: &Hit) -> Spd {
        Spd::default()
    }
}
